File: src/deadlock_build_sync/purchase_decisions.py
"""Group all eligible routes, keeping upgrade components out of PICK ONE lists."""

from __future__ import annotations

from collections import defaultdict
from typing import TYPE_CHECKING

from .purchase_guidance_types import PurchaseDecision

if TYPE_CHECKING:
    from .mechanics import ItemGraph
    from .purchase_guidance_types import PurchaseChoice


def _take_fork(
    seed: PurchaseChoice, remaining: list[PurchaseChoice]
) -> tuple[list[PurchaseChoice], list[PurchaseChoice]]:
    group = [seed]
    parts = set(seed.route[:-1])
    while True:
        joined = [row for row in remaining if parts.intersection(row.route[:-1])]
        if not joined:
            return group, remaining
        group.extend(joined)
        joined_ids = {row.item_id for row in joined}
        remaining = [row for row in remaining if row.item_id not in joined_ids]
        parts.update(item for row in joined for item in row.route[:-1])


def _forks(options: list[PurchaseChoice]) -> list[list[PurchaseChoice]]:
    groups: list[list[PurchaseChoice]] = []
    remaining = list(options)
    while remaining:
        group, remaining = _take_fork(remaining[0], remaining[1:])
        groups.append(group)
    return groups


def _decision(
    position: int, label: str, rows: list[PurchaseChoice], *, fork: bool = False
) -> PurchaseDecision:
    options = tuple(
        row.item_id
        for row in sorted(
            rows,
            key=lambda row: (-(row.timing.buyers if row.timing else 0), row.item_id),
        )
    )
    kind = (
        "PICK ONE"
        if len(rows) > 1
        else "UPGRADE"
        if len(rows[0].route) > 1
        else "OPTIONAL"
    )
    return PurchaseDecision(position, kind, label, options, fork)


def decisions_at(
    position: int, choices: tuple[PurchaseChoice, ...], graph: ItemGraph
) -> tuple[PurchaseDecision, ...]:
    cards = [
        card
        for card in choices
        if card.after_step == position and card.plan is not None
    ]
    ancestors = {item for card in cards for item in card.route[:-1]}
    options = [card for card in cards if card.item_id not in ancestors]
    grouped: dict[str, list[PurchaseChoice]] = defaultdict(list)
    result: list[PurchaseDecision] = []
    for group in _forks(options):
        if len(group) > 1:
            common = set.intersection(*(set(row.route[:-1]) for row in group))
            label = (
                "Upgrade "
                + ", ".join(graph.require(item).name for item in sorted(common))
                if common
                else "Shared component upgrades"
            )
            result.append(_decision(position, label, group, fork=True))
        else:
            card = group[0]
            label = (
                card.purpose.label
                if card.purpose.basis != "unclassified"
                else card.name
            )
            grouped[label].append(card)
    result.extend(_decision(position, label, rows) for label, rows in grouped.items())
    return tuple(sorted(result, key=lambda row: (row.purpose, row.options)))

File: crates/deadlock-guides/src/purchase_decisions.rs
use std::collections::{BTreeMap, BTreeSet};

use deadlock_data::Result;
use deadlock_input::ItemGraph;

use crate::purchase_guidance_types::{PurchaseChoice, PurchaseDecision};

/// # Errors
/// Returns an error when an upgrade references an unknown component.
pub fn build_checkpoint_decisions(
    position: usize,
    choices: &[PurchaseChoice],
    graph: &ItemGraph,
) -> Result<Vec<PurchaseDecision>> {
    let cards = choices
        .iter()
        .filter(|card| card.after_step == Some(position) && card.plan.is_some())
        .collect::<Vec<_>>();
    let ancestors = cards
        .iter()
        .flat_map(|card| components(card))
        .copied()
        .collect::<BTreeSet<_>>();
    let options = cards
        .into_iter()
        .filter(|card| !ancestors.contains(&card.item_id))
        .collect();
    let mut grouped = BTreeMap::<String, Vec<&PurchaseChoice>>::new();
    let mut result = Vec::new();
    for group in group_shared_components(options) {
        let Some(first) = group.first() else {
            continue;
        };
        if group.len() > 1 {
            let mut common = components(first).iter().copied().collect::<BTreeSet<_>>();
            for card in &group[1..] {
                common.retain(|item| components(card).contains(item));
            }
            let label = if common.is_empty() {
                "Shared component upgrades".into()
            } else {
                let names = common
                    .iter()
                    .map(|id| Ok(graph.require(*id)?.name.as_str()))
                    .collect::<Result<Vec<_>>>()?;
                format!("Upgrade {}", names.join(", "))
            };
            result.push(build_decision(position, label, group, true));
        } else {
            let label = if first.purpose.basis == "unclassified" {
                &first.name
            } else {
                &first.purpose.label
            };
            grouped.entry(label.clone()).or_default().push(first);
        }
    }
    result.extend(
        grouped
            .into_iter()
            .map(|(label, rows)| build_decision(position, label, rows, false)),
    );
    result.sort_by(|left, right| {
        left.purpose
            .cmp(&right.purpose)
            .then_with(|| left.options.cmp(&right.options))
    });
    Ok(result)
}

fn components(card: &PurchaseChoice) -> &[u64] {
    card.route
        .split_last()
        .map_or(&[], |(_, components)| components)
}

fn group_shared_components(mut remaining: Vec<&PurchaseChoice>) -> Vec<Vec<&PurchaseChoice>> {
    let mut groups = Vec::new();
    while !remaining.is_empty() {
        let seed = remaining.remove(0);
        let mut group = vec![seed];
        let mut parts = components(seed).iter().copied().collect::<BTreeSet<_>>();
        loop {
            let (joined, rest): (Vec<_>, Vec<_>) = remaining
                .into_iter()
                .partition(|card| components(card).iter().any(|id| parts.contains(id)));
            remaining = rest;
            if joined.is_empty() {
                break;
            }
            parts.extend(joined.iter().flat_map(|card| components(card)).copied());
            group.extend(joined);
        }
        groups.push(group);
    }
    groups
}

fn build_decision(
    position: usize,
    purpose: String,
    mut rows: Vec<&PurchaseChoice>,
    upgrade_fork: bool,
) -> PurchaseDecision {
    rows.sort_by_key(|card| {
        (
            std::cmp::Reverse(card.timing.as_ref().map_or(0, |timing| timing.buyers)),
            card.item_id,
        )
    });
    let kind = if rows.len() > 1 {
        "PICK ONE"
    } else if rows.first().is_some_and(|row| row.route.len() > 1) {
        "UPGRADE"
    } else {
        "OPTIONAL"
    };
    PurchaseDecision {
        after_step: position,
        kind: kind.into(),
        purpose,
        options: rows.iter().map(|row| row.item_id).collect(),
        upgrade_fork,
    }
}
